core/src/main/java/org/ldaptive/LdapAttribute.java

```diff
@@ -591,68 +551,220 @@ public String toString()
 
   /**
    * Returns a new attribute whose values are sorted. String values are sorted naturally. Binary values are sorted using
-   * {@link ByteBuffer#compareTo(ByteBuffer)}.
+   * {@link Arrays#compare(byte[], byte[])}.
    *
    * @param  la  attribute to sort
    *
-   * @return  sorted attribute
+   * @return  new ldap attribute with sorted values
    */
   public static LdapAttribute sort(final LdapAttribute la)
   {
     final LdapAttribute sorted = new LdapAttribute(la.getName());
     if (la.isBinary()) {
       sorted.setBinary(true);
-      final Set<byte[]> newValues = la.getBinaryValues().stream().sorted(
-        (o1, o2) -> {
-          final ByteBuffer bb1 = ByteBuffer.wrap(o1);
-          final ByteBuffer bb2 = ByteBuffer.wrap(o2);
-          return bb1.compareTo(bb2);
-        }).collect(Collectors.toCollection(LinkedHashSet::new));
+      final Set<byte[]> newValues = la.getBinaryValues().stream()
+        .sorted(new ByteArrayComparator()).collect(Collectors.toCollection(LinkedHashSet::new));
       sorted.addBinaryValues(newValues);
     } else {
       final Set<String> newValues = la.getStringValues().stream()
-        .sorted(Comparator.comparing(String::toString)).collect(Collectors.toCollection(LinkedHashSet::new));
+        .sorted(new StringComparator()).collect(Collectors.toCollection(LinkedHashSet::new));
       sorted.addStringValues(newValues);
     }
     return sorted;
   }
 
 
   /**
-   * Converts the supplied string value to a byte array respecting the {@link #binary} flag. If this attribute is
-   * binary, value is expected to be in base64 format. Otherwise, value is UTF-8 encoded.
-   *
-   * @param  value  to convert
-   * @param  throwOnError  whether to throw if a base64 decode error occurs
+   * Creates a builder for this class.
    *
-   * @return  binary value
+   * @return  new builder
+   */
+  public static Builder builder()
+  {
+    return new Builder();
+  }
+
+
+  /**
+   * Container for an attribute value. Supports helpers methods related to byte arrays and the notion of `binary`
+   * attribute values which are base64 encoded.
    *
-   * @throws  IllegalArgumentException  if attribute is binary, value cannot be base64 decoded and throwOnError is true
+   * @author  Middleware Services
    */
-  private byte[] toByteArray(final String value, final boolean throwOnError)
+  private static class AttributeValue
   {
-    if (binary) {
+
+    /**
+     * hash code seed.
+     */
+    private static final int HASH_CODE_SEED = 10723;
+
+    /**
+     * attribute value.
+     */
+    private final byte[] value;
+
+
+    /**
+     * Creates a new attribute value.
+     *
+     * @param bytes byte array value
+     * @param copy  whether to copy the supplied bytes
+     */
+    AttributeValue(final byte[] bytes, final boolean copy)
```

Comment:
I didn't see any usage of this constructor with the copy flag set to true. Did I miss it?

Reply:
Yes, `addValues` and `addBinaryValues` both use it.